미션 제출합니다.

chanhalee opened the pull request.

pobiconan committed:
initial commit

javajigi committed:
docs: 미션 제목 변경

javajigi committed:
add nextstep utils, test code, update library version

javajigi committed:
refactor test code

javajigi committed:
refactor names test data

javajigi committed:
refactor winner message

javajigi committed:
refactor desc to main method

javajigi committed:
update the latest version

javajigi committed:
remove Car class

javajigi committed:
fix(racingcar): correct the test expected text

chanhalee committed:
docs(README) 요구사항 정리

페어와 요구사항을 정리함

chanhalee committed:
build(gradle) Java 버전 및 gradle 버전 수정

Java 17, gradle 7.2로 업데이트

chanhalee committed:
feat(Car) Car 전진 요구사항 테스트 작성 및 구현

자동차가 전진하면 위치가 1 증가한다

chanhalee committed:
docs(README) 요구사항 업데이트

요구사항 구체화

chanhalee committed:
feat(NumberGenerator) 랜덤 숫자 생성기 테스트 및 구현

0에서 9 사이의 정수를 생성하는 숫자 생성기 테스트 제작과 구현

chanhalee committed:
feat(ProceedLogic) 전진 여부를 판정하는 로직 테스트 작성 및 구현

전진 여부를 판정하는 로직 테스트 제작과 구현

chanhalee committed:
feat(RacingGame) 자동차들 중 가장 많이 이동한 자동차가 우승자가 된다.

자동차들 중 가장 많이 이동한 자동차가 우승자가 된다.
- 복수의 우승자가 발생할 수 있다.

chanhalee committed:
docs(README) 진행 현황 업데이트

진행 현황 업데이트

chanhalee committed:
refactor(Cars) 클래스 이름 변경

클래스 역활 일부 수정에 의한 명칭 변경
RacingGame -> Cars

chanhalee committed:
feat(UI) CLI 인터페이스 구현

사용자와 상호작용할 CLI 인터페이스 구현
- CarState dto 클래스 생성

chanhalee committed:
feat(UI) 에러 메시지 출력 ui작성

각 상황별 에러메시지를 출력하는 ui구현

chanhalee committed:
feat(controller) ui와 도메인을 연결하는 컨트롤러 로직 작성

사용자 입력 검증 및 레이싱 게임 실행 로직 구현

chanhalee committed:
docs(README) 프로젝트 진행 현황 업데이트

프로젝트 진행 현황 업데이트

chanhalee committed:
refactor(RacingGame) validation 로직 리팩토링

validation 로직 리팩토링

chanhalee committed:
docs(README) 문자열 덧셈 계산기 TODO 작성

문자열 덧셈 계산기 TODO 작성

chanhalee committed:
feat(StringParser) 문자열 파서 구현

입력된 문자열을 해석해서 정수형 리스트를 반환하는 파서 구현

chanhalee committed:
feat(StringCalculator) 문자열 덧셈 계산기 구현

입력된 문자열을 해석해서 덧셈 연산 결과를 반환하는 계산기 구현

chanhalee committed:
refactor(StringParsor) 파서 리팩토링

코드 컨벤션을 준수하도록 리팩토링

chanhalee committed:
fix: git repository different conflict

hyunssooo commented on build.gradle:
```diff
@@ -2,20 +2,21 @@ plugins {
 dependencies {
-    testImplementation 'org.assertj:assertj-core:3.22.0'
-    testImplementation 'org.junit.jupiter:junit-jupiter:5.8.2'
+    implementation 'com.github.woowacourse-projects:mission-utils:1.0.0'
 }
 
 java {
     toolchain {
-        languageVersion = JavaLanguageVersion.of(11)
+        languageVersion = JavaLanguageVersion.of(17)
     }
```
이 부분은 따로 가이드가 있었을까요??

chanhalee replied:
없었습니다.
수정하도록 하겠습니다.

hyunssooo commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,9 @@
 distributionUrl=https\://services.gradle.org/distributions/gradle-7.5.1-bin.zip
+=======
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.2-bin.zip
+>>>>>>> precourse/chanhalee
```
충돌나는데 괜찮나요?

hyunssooo commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,47 @@
+public class Car implements Comparable<Car>{
+
+	private final String name;
+	private int position;
```
전체적으로 tab을 사용하고 계신 데 보통 space 4칸을 많이 사용합니다. 코드 뷰어마다 다르게 보이기 때문이에요. 깃허브에서만 봐도 idea와는 다르게 보일 거예요. space 4칸으로 맞춰 주시면 감사합니다!

hyunssooo commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,47 @@
+		return Objects.hash(name, position);
+	}
+
+	@Override
+	public int compareTo(Car other) {
+		return other.position - this.position;
+	}
```
compareTo 👍 
`Integer#compare` 를 사용해도 좋을 것 같아요. 

```java
    @Override
    public int compareTo(Car other) {
        return Integer.compare(this.position, other.position);
    }
```

hyunssooo commented on src/main/java/stringcalculator/StringParser.java:
```diff
@@ -0,0 +1,49 @@
+	public static List<Integer> parse(String input) {
+		String delimiter = DEFAULT_DELIMITER;
+		Matcher matcher = Pattern.compile("//(.)\n(.*)").matcher(input);
+		if (matcher.find()) {
```
`Pattern.compile("//(.)\n(.*)")` 이 부분은 한번 생성하고 재사용 가능하지 않을까요?
이펙티브 자바 아이템 6. 불필요한 객체 생성을 피하라 를 읽어보시면 좋을 것 같아요.

hyunssooo commented on src/test/java/racingcar/CarTest.java:
```diff
@@ -0,0 +1,18 @@
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class CarTest {
+
```
car 이름에 대한 테스트는 필요하지 않을까요?
`name` 이라는 객체가 있다면 위임할 수 도 있을 것 같아요!

chanhalee replied:
name이라는 객체가 어떤 것을 의미하는 것인지 잘 모르겠습니다. 조금만 더 자세히 말씀해주시면 감사하겠습니다.

Car 내부에서 이름 중복 검사가 이루어질 수 없다고 생각하여 RacingGame에 Car List를 보관하고 Racing 게임에서 이름의 중복, 글자 수에 대한 validation 로직을 구현했습니다.
하지만 또 validation 메서드를 외부에 공유할 이유가 없다 보니 private로 구현하는 것이 맞다고 생각했습니다.
이름에 대한 검증도 중요하지만 그렇다고 그 검증을 위해 메서드 접근제어자를 바꾸는 것은 적절하지 않는 듯하여 private로 구현하고 테스트 로직을 넣지 않았습니다.
다른 방식으로 바꾸는 게 좋을까요?

hyunssooo replied:
Car.name은 현재 String인데 Name이라는 클래스를 작성하고 Car에서는 Name 객체를 필드에 가지고 있다면 이름에 대한 제약 조건(5자이하) 에 대한 책임을 Name 객체가 가질 수 있다고 생각했어요. :) 중복에 대한 제약 조건은 요구 사항에 보이지 않네요. 

> validation 메서드를 외부에 공유할 이유가 없다 보니 private로 구현하는 것이 맞다고 생각했습니다.

저도 동의합니다. 하지만 현재 Car 클래스 자체만 보면 name의 길이는 상관없이 생성되고 있어요. 

```java
Car car = new Car("aaaaaa"); // 생성 가능
```

객체 지향에서 객체들의 협력으로 요구 사항을 풀어냅니다. 저는 `이름은 5자 이하만 가능하다.` 는 비즈니스 요구 사항이고 이는 모델내에 위치해야 한다고 생각해요.

그렇기 때문에 저는 Car에 대한 생성에 name에 대한 테스트가 필요하다고 생각했어요. 생각하시는 비즈니스 로직은 어디에 위치해야 하며 도메인 모델에 위치하는 로직과 도메인 모델 밖에 위치하는 로직의 기준은 무엇일까요?

chanhalee replied:
생각해보니 말씀하신 것처럼 Car에 이름 생성에 대한 로직을 넣는 것이 좋을 것 같습니다. 
수정하도록 하겠습니다.

hyunssooo commented on src/test/java/stringcalculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,32 @@
+
+public class StringCalculatorTest {
+
+	@Test
+	@DisplayName("문자열을 입력받아 연산 결과를 반환")
+	public void calculateTest() {
+		String input = "1,2,3";
+		Assertions.assertThat(StringCalculator.calculate(input)).isEqualTo(6);
+
+		input = "1:2:3";
+		Assertions.assertThat(StringCalculator.calculate(input)).isEqualTo(6);
+
+		input = "1:2,3";
+		Assertions.assertThat(StringCalculator.calculate(input)).isEqualTo(6);
+
+		input = "1,2:3,10";
+		Assertions.assertThat(StringCalculator.calculate(input)).isEqualTo(16);
+
+		input = "1,,3";
+		Assertions.assertThat(StringCalculator.calculate(input)).isEqualTo(4);
+
+		input = "1, ,3";
+		Assertions.assertThat(StringCalculator.calculate(input)).isEqualTo(4);
+	}
```
코드가 반복되고 있는데요. 중복을 제거해 보아요.
`@ParameterizedTest`, `@ValueSource` 를 찾아보고 적용해 보면 좋을 것 같아요.

hyunssooo replied:
https://edu.nextstep.camp/s/MCLQmhAp/ls/TB7zBs5c 의 
기능 요구 사항은 모두 테스트로 작성되면 좋을 것 같습니다. 테스트가 빈약해 보여요.

chanhalee replied:
private으로 접근 제한된 메서드에서 기능이 구현된 것들이 있습니다.
(예를들어 RacingGame에서 findWinner, validation, ...)
이들을 테스트하기 위해 public으로 변경하는 것이 좋을까요?

hyunssooo commented on src/test/java/stringcalculator/StringParserTest.java:
```diff
@@ -0,0 +1,77 @@
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class StringParserTest {
```
여기도 중복이 많이 보이네요!

hyunssooo commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+import java.util.Set;
+
+import racingcar.dto.CarState;
+
+public class RacingGame {
+
```
보통 위에서 아래로 코드를 읽어가는데요. 메서드 순서도 고민해 보시면 좋을 것 같아요.

hyunssooo commented on src/main/java/racingcar/RacingGame.java:
```diff
@@ -0,0 +1,141 @@
+		rounds = Integer.parseInt(input);
+	}
+
+	private void initCars() {
+		boolean validationPass = false;
+		List<String> carNameList = null;
```
carNameList 보다는 복수형인 carNames 는 어떨까요? 관심사가 자료구조인 List가 아닌 내부의 복수 CarName이기 때문이에요.

hyunssooo reviewed (changes requested):
안녕하세요. 🙂
이번 미션 함께 진행할 정현수입니다.
몇 가지 코멘트 남겼어요. 확인 부탁드립니다!

chanhalee committed:
refactor (PR 코멘트 1번,2번에 대한 수정)

PR 코멘트 1번,2번에 대한 수정

chanhalee committed:
refactor (PR 코멘트 5번 대응 수정)

PR 코멘트 5번에 대응 수정
- Pattern 객체를 1번만 생성하도록 static 변수화

chanhalee committed:
refactor (PR 코멘트 6번 대응 수정)

PR 코멘트 6번에 대응 수정
- Car 생성시 이름 글자수를 체크하는 Test code 생성
- Car 생성자에 이름에 대한 검수 로직 구현

chanhalee committed:
refactor (PR 코멘트 7, 8, 9, 10번 대응 수정)

PR 코멘트 7번에 대응 수정
- 테스트 코드 내 중복 제거
- RacingGame의 findWinner 메서드 테스트 코드 추가
- 변수명에 자료형이 들어가지 않도록 수정
- RacingGame 내 메서드 순서 변경

chanhalee commented:
코멘트를 참고해 코드 수정을 완료했습니다.

hyunssooo commented on src/test/java/stringcalculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,20 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class StringCalculatorTest {
+
```
https://github.com/next-step/java-racingcar-kakao/pull/22#discussion_r1540612037
전체적으로 확인해 보면 좋을 것 같습니다! 
idea에 설정하는 방법이 있는데요. (tab을 눌려도 space 4칸 입력, code style) 
이 부분을 찾아보시면 좋을 것 같아요.

chanhalee replied:
탭에 관련해서
![tabs-spaces](https://github.com/next-step/java-racingcar-kakao/assets/77485580/fb92b88c-1fae-4868-99b6-c5552cf503fc)

사전과제에서 받은 컨벤션 규칙에는 하드탭을 사용하라고 되어있습니다.
4.1. 하드탭 사용
[indentation-tab]
탭(tab) 문자를 사용하여 들여쓴다. 탭 대신 스페이스를 사용하지 않는다. 이를 잘 준수하기 위해서 스페이스와 탭을 구별해서 보여주도록 에디터를 설정한다.
4.2. 탭의 크기는 4개의 스페이스
[4-spaces-tab]
1개의 탭의 크기는 스페이스 4개와 같도록 에디터에서 설정한다.
(출처: https://naver.github.io/hackday-conventions-java/#indentation-tab)

반면에 교육 홈페이지에 나온 다른 2종류의 컨벤션에는 다음과 같이 나와 있습니다.
4. 들여쓰기
4개의 빈 칸(space)를 들여쓰기 단위로 사용한다. 들여쓰기의 정확한 구현(빈 칸을 사용하거나 탭을 사용하거나)은 정해져 있지 않다. 탭은 4개의 빈 칸이 아니라, 8개의 빈 칸으로 설정하는 것이 좋다.
(출처: https://myeonguni.tistory.com/1596)

Whitespace characters
Aside from the line terminator sequence, the ASCII horizontal space character (0x20) is the only whitespace character that appears anywhere in a source file. This implies that:
All other whitespace characters in string and character literals are escaped.
Tab characters are not used for indentation.
(출처: https://google.github.io/styleguide/javaguide.html#s2.3.1-whitespace-characters)

----

저는 하드탭(4space 크기)을 사용하고 깃헙 Appearance 설정을 통해 4space로 설정하여 사용하는 편입니다.
혹시 별로 문제가 안된다면 tab을 사용해도 될까요?

hyunssooo commented on src/test/java/stringcalculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,20 @@
+
+public class StringCalculatorTest {
+
+	@ParameterizedTest
+	@DisplayName("문자열을 입력받아 연산 결과를 반환")
+	@CsvSource(value = {"1,2,3|6","1:2:3|6","1:2,3|6","1,2:3,10|16", "1,,3|4","1, ,3|4"}, delimiter = '|')
+	public void calculateTest(String input, int expected) {
+		Assertions.assertThat(StringCalculator.calculate(input)).isEqualTo(expected);
+	}
```
중복 제거 👍 
다만 테스트가 빈약해 보이는데요. 

> 빈 문자열 또는 null 값을 입력할 경우 0을 반환해야 한다.(예 : “” => 0, null => 0)
> 숫자 하나를 문자열로 입력할 경우 해당 숫자를 반환한다.(예 : “1”)

등의 요구 사항은 테스트로 검증되면 좋지 않을까요?? 

https://github.com/next-step/java-racingcar-kakao/pull/22#discussion_r1545671584

private 메서드 테스트를 위해 public 으로 변경하는게 좋은 방법일까요? 
개인적으로는 선호하지 않는데요. 저는 테스트 대상은 해당 객체의 인터페이스라고 생각합니다. 자바의 interface와는 다르고 보통 public 메서드들인데요. `CarTest#carCreateFailTestFail` 와 같이 public 메서드들을 통해서 private 메서드들도 테스트 가능하지 않을까요?

chanhalee replied:
테스트 보강하겠습니다.

hyunssooo reviewed (approved):
안녕하세요. 
몇 가지 추가 코멘트 남겼어요. 
다음 step 진행하시면서 확인 부탁드려요! 
다음 step을 위해서 이만 머지하겠습니다!
